interfaces/apparmor/backend.go

```diff
@@ -0,0 +1,144 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor implements integration between snappy and
+// ubuntu-core-launcher around apparmor.
+//
+// Snappy creates apparmor profiles for each application (for each snap)
+// present in the system.  Upon each execution of ubuntu-core-launcher, the
+// profile is attached to the running process. Prior to that the profile must
+// be parsed, compiled and loaded into the kernel using the support tool
+// "apparmor_parser".
+//
+// Each apparmor profile contains a simple <header><content><footer> structure.
+// The header specified an identifier that is relevant to the kernel. The
+// identifier can be either the full path of the executable or an abstract
+// identifier not related to the executable name.
+//
+// The actual profiles are stored in /var/lib/snappy/apparmor/profiles.
+// This directory is also hard-coded in ubuntu-core-launcher.
+//
+// NOTE: A systemd job (TODO: specify which) loads all snappy-specific apparmor
+// profiles into the kernel during the boot process.
+package apparmor
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/interfaces"
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+// Backend is responsible for maintaining apparmor profiles for ubuntu-core-launcher.
+type Backend struct {
+	// CustomTemplate exists to support old-security which goes
+	// beyond what is possible with pure security snippets.
+	//
+	// If non-empty then it overrides the built-in template.
+	CustomTemplate string
+}
+
+// Configure creates and loads apparmor security profiles specific to a given
+// snap. The snap can be in developer mode to make security violations
+// non-fatal to the offending application process.
+//
+// This method should be called after changing plug, slots, connections between
+// them or application present in the snap.
+func (b *Backend) Configure(snapInfo *snap.Info, developerMode bool, repo *interfaces.Repository) error {
+	// Get the snippets that apply to this snap
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name, interfaces.SecurityAppArmor)
+	if err != nil {
+		return fmt.Errorf("cannot obtain security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	// Get the files that this snap should have
+	content, err := b.combineSnippets(snapInfo, developerMode, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapInfo.Name, err)
+	}
+	glob := interfaces.SecurityTagGlob(snapInfo)
+	changed, removed, err := osutil.EnsureDirState(dirs.SnapAppArmorDir, glob, content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
+	err = b.observeChanges(changed, removed)
```

Comment:
@zyga that was strawman, but observeChanges is really not ideal as a name, also why not have two functions one for the unload part and one for the load/reload